lightning_examples/finetuning-scheduler/finetuning-scheduler.py

```diff
@@ -0,0 +1,565 @@
+        scheduler = {
+            "scheduler": instantiate_registered_class(args=optimizer, init=self.lr_scheduler_init),
+            **self.pl_lrs_cfg,
+        }
+        return [optimizer], [scheduler]
+
+    def configure_callbacks(self):
+        found_fts = [c for c in self.trainer.callbacks if isinstance(c, FinetuningScheduler)]  # type: ignore # noqa
+        if found_fts:
+            self.finetuningscheduler_callback = found_fts[0]
+        return super().configure_callbacks()
+
+
+# %%
+# let's create a finetuning schedule for our model and run an explicitly scheduled finetuning training scenario with it
+# Please see the documentation for a full description of the schedule format
+ft_schedule_yaml = """
+0:
+  params: # the parameters for each phase definition can be fully specified
+  - model.classifier.bias
+  - model.classifier.weight
+1:
+  params: # or specified via a regex
+  - model.albert.pooler.*
+2:
+  params:
+  - model.albert.encoder.*.ffn_output.*
+"""
+ft_schedule_name = "RteBoolqModule_ft_schedule_albert_base.yaml"
+with open(ft_schedule_name, "w") as f:
+    f.write(ft_schedule_yaml)
+
+# %%
+datasets.set_progress_bar_enabled(False)
+pl.seed_everything(42)
+dm = RteBoolqDataModule(model_name_or_path="albert-base-v2", tokenizers_parallelism=False)
+dm.setup("fit")
+optimizer_init = {
+    "class_path": "torch.optim.AdamW",
+    "init_args": {"weight_decay": 1e-05, "eps": 1e-07, "lr": 1e-05},
+}
+lr_scheduler_init = {
+    "class_path": "torch.optim.lr_scheduler.CosineAnnealingWarmRestarts",
+    "init_args": {"T_0": 1, "T_mult": 2, "eta_min": 1e-07},
+}
+pl_lrs_cfg = {"interval": "epoch", "frequency": 1, "name": "CosineAnnealingWarmRestarts"}
+
+model = RteBoolqModule(
+    model_name_or_path="albert-base-v2",
+    optimizer_init=optimizer_init,
+    lr_scheduler_init=lr_scheduler_init,
+    pl_lrs_cfg=pl_lrs_cfg,
+    experiment_tag="fts_explicit",
+)
+callbacks = [
+    FinetuningScheduler(ft_schedule=ft_schedule_name, max_depth=2),  # type: ignore # noqa
+    FTSEarlyStopping(monitor="val_loss", min_delta=0.001, patience=2),  # type: ignore # noqa
+    FTSCheckpoint(monitor="val_loss", save_top_k=5),  # type: ignore # noqa
+]
+example_logdir = "lightning_logs"
+logger = TensorBoardLogger(example_logdir, name="fts_explicit")
+
+# %%
+# optionally start tensorboard and monitor progress graphically while viewing multi-phase finetuning specific training
+# logs in the cell output below by uncommenting the next 3 lines
+# # !mkdir -p $example_logdir
+# # %load_ext tensorboard
+# # %tensorboard --logdir example_logdir
+# disable progress bar by default to focus on multi-phase training logs. Set to True to re-enable if desired
+show_progress = False
+
+# %%
+trainer = pl.Trainer(
+    enable_progress_bar=show_progress,
+    precision=16,
+    gpus=AVAIL_GPUS,
+    callbacks=callbacks,
+    logger=logger,
+)
+trainer.fit(model, datamodule=dm)
```

Comment:
Did you do a full training? We could upload the weights and curve to s3. Would be neat to see the naive fine-tuning vs the tailored on in terms of performance on SuperGLUE. 
Right now, I learned how to use the scheduler, but I don't see its benefits.

Reply:
Yes, it is a full training, though only on a small model (albert-base), but I'm definitely happy to upload the produced checkpoint.

I've got a full comparison of "nofts_baseline" vs "implicit_mode" vs "explicit_mode" finetuning scenarios (w/ identical parameters other than thaw schedule) available [on tensorboard.dev](https://tensorboard.dev/experiment/Qy917MVDRlmkx31A895CzA/#scalars&_smoothingWeight=0&runSelectionState=eyJmdHNfZXhwbGljaXQiOnRydWUsImZ0c19pbXBsaWNpdCI6dHJ1ZSwibm9mdHNfYmFzZWxpbmUiOnRydWV9) and documented in RST in the current [``finetuning_scheduler_module`` documentation](https://github.com/speediedan/pytorch-lightning/blob/finetuning_scheduler_module/docs/source/advanced/finetuning_scheduler.rst#example-scheduled-finetuning-for-superglue). I'm planning to include the comparison of scenarios in whatever form the new user-registered module documentation takes.
![image](https://user-images.githubusercontent.com/7462936/144520056-19d73498-4831-4b45-b3d8-e238a8393ab1.png)
![image](https://user-images.githubusercontent.com/7462936/144520108-f272927f-c171-4f35-a4d4-13d6834a3fbe.png)

I included a "nofts_baseline" vs "implicit_mode" vs "explicit_mode" comparison in an earlier version of this notebook but felt like the additional training session executions consumed too much compute and that it might be clearer just to demonstrate the primary usage pattern of FinetuningScheduler (an explicit user-defined schedule).

Maybe linking in the example notebook to the latest tensorboard.dev comparison of the scenarios would be useful? I can go back to including training all three scenarios in the notebook if you prefer but it's a lot IMHO. 

As far as uploading the weights for experimentation, can I just drop them on google drive? I see the awesome UvA tutorials use github for saved_models but I think there are some limitations to posting the weights there that make google drive preferable.

Thanks for being generous with your time and all the great feedback!

Reply:
Hey @speediedan,

Thanks for the hard work. This is starting to look very great!

I would prefer to actually include the 3 runs with their logs and the checkpoints. This brings much more value to your presentation as readers would be able to relate to the actual experimentations.

So, let's upload the weights and logs. You can use Google Drive or we can upload them to our s3 bucket. It is up to you :) 

Furthermore, it would be great to see an inference using trainer.predict on some examples using all 3 models, especially if the latest creates slightly better predictions !

Getting close to the finish line :)

Reply:
In addition to the recommended changes, I've made a number of expository enhancements and improvements to reproduciblity (in the full DDP Lightning CLI version to keep the tutorial as simple as possible). e.g. you can see salient library versions logged in the tensorboard summary hparams:
![image](https://user-images.githubusercontent.com/7462936/145287600-3f1406a6-8ef6-466c-8bd8-cd4b7cefe62f.png)

Updates in summary:
1. Added ``nofts_baseline`` and ``fts_implicit`` mode scenarios with full [tensorboard experiment summaries](https://tensorboard.dev/experiment/n7U8XhrzRbmvVzC4SQSpWw/), [log files](https://drive.google.com/file/d/1LrUcisRLHeJgh_BDOOD_GUBPp5iHAkoR/view?usp=sharing) and [associated checkpoints](https://drive.google.com/file/d/1t7myBgcqcZ9ax_IT9QVk-vFH_l_o5UXB/view?usp=sharing). I think it makes sense to move the checkpoints to the PL s3 bucket since I believe the bandwidth quotas on google drive will be lower than your s3 bucket (not that we will be in danger of hitting it, but just in case). 
2. Clarified the value conferred by using ``FinetuningScheduler`` and demonstrated some of the model exploration utility
3. Switched to using the recently release DeBERTav3 model from microsoft (using the smaller base version to allow running on modest GPUs)

Though the performance on RTE with the ``fts_explicit`` scenario is on the order of a few % better than the naive ``nofts_baseline`` and a few examples could be found that the explicit model classifies correctly that the ``nofts_baseline`` scenario doesn't, I don't think show those few classification examples will be especially illuminating. Though you can squeeze out a few additional percentage points of performance with the ``FinetuningScheduler`` in many contexts, I would prefer to keep the focus on the model research/exploration benefits if that's okay with you. Hope you like the improvements! I know there will likely be some additional changes but I know we're getting there. Thanks again for your all your work and leadership!

Reply:
Hey @speediedan, Yes, that's entirely fine and I think it is better to keep your vision in this notebook.